src/cider/exec/nextgen/function/CiderDateFunctions.cpp

```diff
@@ -0,0 +1,282 @@
+constexpr unsigned SUNDAY = 3;
+constexpr unsigned SATURDAY = 4;
+
+// Clamp day-of-month to max day of the month. E.g. April 31 -> 30.
+static unsigned clampDom(unsigned yoe, unsigned moy, unsigned dom) {
+  constexpr unsigned max_days[11]{30, 29, 30, 29, 30, 30, 29, 30, 29, 30, 30};
+  if (dom < 28) {
+    return dom;
+  } else {
+    unsigned const max_day = moy == 11
+                                 ? 27 + (++yoe % 4 == 0 && (yoe % 100 != 0 || yoe == 400))
+                                 : max_days[moy];
+    return dom < max_day ? dom : max_day;
+  }
+}
+
+// If OFFSET=MONDAY,
+// then return day-of-era of the Monday of ISO 8601 week 1 in the given year-of-era.
+// Similarly for SUNDAY and SATURDAY. In all cases, week 1 always contains Jan 4.
+template <unsigned OFFSET>
+unsigned week_start_from_yoe(unsigned const yoe) {
+  unsigned const march1 = yoe * 365 + yoe / 4 - yoe / 100;
+  unsigned const jan4 = march1 + (MARJAN + 3);
+  unsigned const jan4dow = (jan4 + OFFSET) % 7;
+  return jan4 - jan4dow;
+}
+
+class MonthDaySecond {
+  int64_t months_;  // Number of months since 2000 March 1.
+  int64_t era_;
+  OffsetUnit offset_unit_;
+
+ public:
+  MonthDaySecond(int32_t const date, unsigned sod = 0) {
+    offset_unit_.sod = sod;
+    era_ = floor_div(date - kEpochAdjustedDays, kDaysPer400Years);
+    offset_unit_.doe = date - kEpochAdjustedDays - era_ * kDaysPer400Years;
+    offset_unit_.yoe = (offset_unit_.doe - offset_unit_.doe / 1460 +
+                        offset_unit_.doe / 36524 - offset_unit_.doe / 146096) /
+                       365;
+    offset_unit_.doy = offset_unit_.doe - (365 * offset_unit_.yoe + offset_unit_.yoe / 4 -
+                                           offset_unit_.yoe / 100);
+    offset_unit_.moy = (5 * offset_unit_.doy + 2) / 153;
+    offset_unit_.dom = offset_unit_.doy - (153 * offset_unit_.moy + 2) / 5;
+    months_ = (era_ * 400 + offset_unit_.yoe) * 12 + offset_unit_.moy;
+  }
+
+  MonthDaySecond& addMonths(int64_t const months) {
+    months_ += months;
+    updateOffset();
+    return *this;
+  }
+
+  void updateOffset() {
+    era_ = floor_div(months_, 12 * 400);
+    offset_unit_.moe = months_ - era_ * (12 * 400);
+    offset_unit_.yoe = offset_unit_.moe / 12;
+    offset_unit_.moy = offset_unit_.moe % 12;
+    offset_unit_.doy = (153 * offset_unit_.moy + 2) / 5 +
+                       clampDom(offset_unit_.yoe, offset_unit_.moy, offset_unit_.dom);
+    offset_unit_.doe = offset_unit_.yoe * 365 + offset_unit_.yoe / 4 -
+                       offset_unit_.yoe / 100 + offset_unit_.doy;
+  }
+
+  // Return number of seconds since 1 January 1970.
+  int64_t getTime() const { return getDate() * kSecsPerDay + offset_unit_.sod; }
+
+  // Return dates since 1 January 1970.
+  int32_t getDate() const {
+    return (kEpochAdjustedDays + era_ * kDaysPer400Years + offset_unit_.doe);
+  }
+
+  int64_t extractYear() const {
+    return 2000 + era_ * 400 + offset_unit_.yoe + (MARJAN <= offset_unit_.doy);
```

Comment:
only support years after 2000?

Reply:
era_ will be a negative value if date is before 2000.